Sources/GRPCCodeGen/Internal/Translator/IDLToStructuredSwiftTranslator.swift

```diff
@@ -49,6 +50,87 @@ struct IDLToStructuredSwiftTranslator: Translator {
   }
 }
 
+extension IDLToStructuredSwiftTranslator {
+  private func validateInput(_ codeGenerationRequest: CodeGenerationRequest) throws {
+    let servicesByNamespace = Dictionary(
+      grouping: codeGenerationRequest.services,
+      by: { $0.namespace }
+    )
+    try checkServiceNamesAreUnique(for: servicesByNamespace)
+    for service in codeGenerationRequest.services {
+      try checkMethodNamesAreUnique(in: service)
+    }
+  }
+
+  // Verify service names are unique within each namespace and that services with no namespace
+  // don't have the same names as any of the namespaces.
+  private func checkServiceNamesAreUnique(
+    for servicesByNamespace: [String: [CodeGenerationRequest.ServiceDescriptor]]
+  ) throws {
+    // Check that if there are services in an empty namespace, none have names which match other namespaces
+    let noNamespaceServices = servicesByNamespace["", default: []]
+    let namespaces = servicesByNamespace.keys
+    for service in noNamespaceServices {
+      if namespaces.contains(service.name) {
+        throw CodeGenError(
+          code: .nonUniqueServiceName,
+          message: """
+            Services with no namespace must not have the same names as the namespaces. \
+            \(service.name) is used as a name for a service with no namespace and a namespace.
+            """
+        )
+      }
+    }
```

Comment:
Very minor nit, but you can skip the `, default: []]` subscript and instead make this whole block conditional to it being an empty key:

```suggestion
    if let noNamespaceServices = servicesByNamespace[""] {
        let namespaces = servicesByNamespace.keys
        for service in noNamespaceServices {
          if namespaces.contains(service.name) {
            throw CodeGenError(
              code: .nonUniqueServiceName,
              message: """
                Services with no namespace must not have the same names as the namespaces. \
                \(service.name) is used as a name for a service with no namespace and a namespace.
                """
            )
          }
      }
    }
```